cmd/habitat/node/node.go

```diff
@@ -0,0 +1,105 @@
+package node
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+	"os/signal"
+
+	"github.com/eagraf/habitat/cmd/habitat/api"
+	dataproxy "github.com/eagraf/habitat/cmd/habitat/data_proxy"
+	"github.com/eagraf/habitat/cmd/habitat/procs"
+	"github.com/eagraf/habitat/cmd/habitat/proxy"
+	"github.com/eagraf/habitat/cmd/sources"
+	"github.com/eagraf/habitat/pkg/compass"
+	"github.com/eagraf/habitat/pkg/p2p"
+	"github.com/eagraf/habitat/structs/ctl"
+	ma "github.com/multiformats/go-multiaddr"
+	"github.com/rs/zerolog/log"
+	"github.com/spf13/viper"
+)
+
+const (
+	ReverseProxyHost = "0.0.0.0"
+	ReverseProxyPort = "2041"
+
+	P2PPort = "6000"
+)
+
+type Node struct {
+	ID string
+
+	ProcessManager *procs.Manager
+	P2PNode        *p2p.Node
+	ReverseProxy   *proxy.Server
+	DataProxy      *dataproxy.DataProxy
+}
+
+func NewNode() (*Node, error) {
+	priv, _ := compass.GetPeerIDKeyPair()
+
+	p2pNode, err := p2p.NewNode(P2PPort, priv)
+	if err != nil {
+		return nil, err
+	}
+	log.Info().Msgf("starting LibP2P node with peer ID %s listening at port %s", p2pNode.Host().ID().Pretty(), P2PPort)
+
+	procsDir := compass.ProcsPath()
+	reverseProxy := proxy.NewServer()
+
+	return &Node{
+		ID:             compass.NodeID(),
+		P2PNode:        p2pNode,
+		ReverseProxy:   reverseProxy,
+		DataProxy:      dataproxy.NewDataProxy(map[string]*sources.DataServerNode{}),
+		ProcessManager: procs.NewManager(procsDir, reverseProxy.Rules),
+	}, nil
+}
+
+func (n *Node) Start() error {
+	// Start reverse proxy
+	proxyAddr := fmt.Sprintf("%s:%s", ReverseProxyHost, ReverseProxyPort)
+	go n.ReverseProxy.Start(proxyAddr)
+
+	redirectURL, err := url.Parse("http://" + proxyAddr + "/habitat")
+	if err != nil {
+		log.Fatal().Err(err)
+	}
+	go proxy.LibP2PHTTPProxy(n.P2PNode.Host(), redirectURL)
+
+	// Start data proxy
+	sourcesPort := viper.GetString("data-proxy-port")
+	go n.DataProxy.Start(context.Background(), sourcesPort)
+
+	// Start process manager
+	go n.ProcessManager.ListenForErrors()
+	go handleInterupt(n.ProcessManager)
+
+	// Keep this thread running
+	select {}
```

Comment:
not a big deal but instead of `select {}` i think the conventional way to do this is have a context we pass into each of the process and wait at the end for `<-ctx.Done()`

Reply:
Oh yeah thats a lot nicer